linear_operator/operators/kernel_linear_operator.py

```diff
@@ -0,0 +1,217 @@
+    so that :math:`\boldsymbol K_{ij} = k_{\boldsymbol \theta}([\boldsymbol X_1]_i, [\boldsymbol X_2]_j)`.
+
+    The output of :math:`k_{\boldsymbol \theta}(\cdot,\cdot)` (`covar_func`) can either be a torch.Tensor
+    or a LinearOperator.
+
+    .. note ::
+
+        Each of the passed parameters should have a minimum of 2 (likely singleton) dimensions.
+        Any additional dimension will be considered a batch dimension that broadcasts with the batch dimensions
+        of x1 and x2.
+
+        For example, to implement the RBF kernel
+
+        .. math::
+
+            o^2 \exp\left(
+                -\tfrac{1}{2} (\boldsymbol x_1 - \boldsymbol x2)^\top \boldsymbol D_\ell^{-2}
+                (\boldsymbol x_1 - \boldsymbol x2)
+            \right),
+
+        where :math:`o` is an `outputscale` parameter and :math:`D_\ell` is a diagonal `lengthscale` matrix,
+        we would expect the following shapes:
+
+        - `x1`: `(*batch_shape x N x D)`
+        - `x2`: `(*batch_shape x M x D)`
+        - `lengthscale`: `(*batch_shape x 1 x D)`
+        - `outputscale`: `(*batch_shape x 1 x 1)`
+
+        Not adding the appropriate singleton dimensions to `lengthscale` and `outputscale` will lead to erroneous
+        kernel matrix outputs.
+
+    .. code-block:: python
+
+        # NOTE: _covar_func intentionally does not close over any parameters
+        def _covar_func(x1, x2, lengthscale, outputscale):
+            # RBF kernel function
+            # x1: ... x N x D
+            # x2: ... x M x D
+            # lengthscale: ... x 1 x D
+            # outputscale: ... x 1 x 1
+            x1 = x1.div(lengthscale)
+            x2 = x2.div(lengthscale)
+            sq_dist = (x1.unsqueeze(-2) - x2.unsqueeze(-3)).square().sum(dim=-1)
+            kern = sq_dist.div(-2.0).exp().mul(outputscale.square())
+            return kern
+
+
+        # Batches of data
+        x1 = torch.randn(3, 5, 6)
+        x2 = torch.randn(3, 4, 6)
+        # Broadcasting lengthscale and output parameters
+        lengthscale = torch.randn(2, 1, 1, 6)
+        outputscale = torch.randn(2, 1, 1, 1)
+        kern = KernelLinearOperator(x1, x2, lengthscale, outputscale, covar_func=covar_func)
+
+        # kern is of size 2 x 3 x 5 x 4
+
+    .. warning ::
+
+        `covar_func` should not close over any parameters. Any parameters that are closed over will not have
+        propagated gradients.
+
+    :param x1: The data :math:`\boldsymbol X_1.`
+    :param x2: The data :math:`\boldsymbol X_2.`
+    :param params: Additional hyperparameters (:math:`\boldsymbol \theta`) passed into covar_func.
+    :param covar_func: The covariance function :math:`k_{\boldsymbol \theta}(\cdot, \cdot)`.
+        Its arguments should be `x1`, `x2`, `*params`, `**kwargs`, and it should output the covariance matrix
+        between :math:`\boldsymbol X_1` and :math:`\boldsymbol X_2`.
+    :param kwargs: Any additional (non-hyperparameter) kwargs to pass into `covar_func`.
```

Comment:
It's a bit weird that all hypers are assumed to be positional args here. It seems that this is rather opaque gotcha, and could also result in making the code much less readable.

Reply:
@Balandat I can try to allow for kwarg hypers, but this might be a can of worms.

Reply:
@Balandat okay I changed it so that hypers must be kwargs (cannot be positional args).

This change required modifying some base LinearOperator functionality so that the flattening/unflattening that we do in representation trees (for CG backward passes, etc.) accounts for LinearOperator kwargs as well as args.